File: v1/graphs/reducers/inventory.js
// Deterministic repository inventory: source files grouped into modules (top-level dirs), sizes, languages.
// input: { repo_path }   args: { include?: ["src","graphs"], max_files?: 400, exts?: [".ts",".js"], module_depth?: 2 }
import fs from "node:fs";
import path from "node:path";

const IGNORE = new Set(["node_modules", "dist", ".git", "runs", "out", ".claude", "coverage", "build", "target", "__pycache__"]);

export default async function inventory(input, args, ctx) {
  const root = path.resolve(String(input.repo_path ?? args.root ?? "."));
  const inc = Array.isArray(input.include) ? input.include : args.include;
  const include = Array.isArray(inc) && inc.length ? inc.map(String) : null;
  const exts = new Set((Array.isArray(args.exts) ? args.exts : [".ts", ".tsx", ".js", ".mjs", ".py", ".go", ".rs", ".java", ".yaml", ".yml", ".json", ".md", ".html"]).map(String));
  const maxFiles = Number(args.max_files ?? 400);
  const depth = Number(args.module_depth ?? 2);
  const files = [];
  const walk = (dir, rel) => {
    if (files.length >= maxFiles) return;
    for (const entry of fs.readdirSync(dir, { withFileTypes: true })) {
      if (IGNORE.has(entry.name) || entry.name.startsWith(".")) continue;
      const abs = path.join(dir, entry.name);
      const r = rel ? `${rel}/${entry.name}` : entry.name;
      if (entry.isDirectory()) walk(abs, r);
      else if (exts.has(path.extname(entry.name))) {
        if (include && !include.some((p) => r === p || r.startsWith(`${p}/`))) continue;
        const text = fs.readFileSync(abs, "utf8");
        files.push({ path: r, bytes: Buffer.byteLength(text), lines: text.split("\n").length, ext: path.extname(entry.name) });
        if (files.length >= maxFiles) return;
      }
    }
  };
  walk(root, "");
  const modules = {};
  for (const f of files) {
    const parts = f.path.split("/");
    const key = parts.length > depth ? parts.slice(0, depth).join("/") : parts.length > 1 ? parts.slice(0, -1).join("/") : "(root)";
    const m = (modules[key] ??= { module: key, files: [], lines: 0, bytes: 0 });
    m.files.push(f.path);
    m.lines += f.lines;
    m.bytes += f.bytes;
  }
  const list = Object.values(modules).sort((a, b) => b.lines - a.lines);
  ctx.log(`inventory: ${files.length} files in ${list.length} modules under ${root}`);
  return {
    root: root.replace(/\\/g, "/"),
    file_count: files.length,
    total_lines: files.reduce((s, f) => s + f.lines, 0),
    modules: list,
    largest_files: [...files].sort((a, b) => b.lines - a.lines).slice(0, 10),
    _stats: { in: files.length, out: list.length },
  };
}

File: graphs/reducers/inventory.py
# Deterministic repository inventory: source files grouped into modules (top-level dirs), sizes, languages.
# input: { repo_path, include? }   args: { include?: ["gren","graphs"], max_files?: 400, exts?: [".py"], module_depth?: 2 }
import os

IGNORE = {"node_modules", "dist", ".git", "runs", "out", ".claude", "coverage", "build", "target", "__pycache__", ".venv", "venv", ".worktrees", ".pytest_cache", "v1"}
DEFAULT_EXTS = [".py", ".ts", ".tsx", ".js", ".mjs", ".go", ".rs", ".java", ".yaml", ".yml", ".json", ".md", ".html", ".toml"]


def reduce(input, args, ctx):
    root = os.path.abspath(str(input.get("repo_path") or args.get("root") or "."))
    inc = input.get("include") if isinstance(input.get("include"), list) else args.get("include")
    include = [str(p) for p in inc] if isinstance(inc, list) and inc else None
    exts = {str(e) for e in (args.get("exts") if isinstance(args.get("exts"), list) else DEFAULT_EXTS)}
    max_files = int(args.get("max_files", 400))
    depth = int(args.get("module_depth", 2))
    files: list[dict] = []

    def walk(d: str, rel: str) -> None:
        if len(files) >= max_files:
            return
        try:
            entries = sorted(os.scandir(d), key=lambda e: e.name)
        except OSError:
            return
        for entry in entries:
            if entry.name in IGNORE or entry.name.startswith("."):
                continue
            r = f"{rel}/{entry.name}" if rel else entry.name
            if entry.is_dir(follow_symlinks=False):
                walk(entry.path, r)
            elif os.path.splitext(entry.name)[1] in exts:
                if include and not any(r == p or r.startswith(f"{p}/") for p in include):
                    continue
                try:
                    with open(entry.path, encoding="utf-8", errors="replace") as f:
                        text = f.read()
                except OSError:
                    continue
                files.append({"path": r, "bytes": len(text.encode("utf-8")), "lines": text.count("\n") + 1, "ext": os.path.splitext(entry.name)[1]})
                if len(files) >= max_files:
                    return

    walk(root, "")
    modules: dict[str, dict] = {}
    for f in files:
        parts = f["path"].split("/")
        key = "/".join(parts[:depth]) if len(parts) > depth else "/".join(parts[:-1]) if len(parts) > 1 else "(root)"
        m = modules.setdefault(key, {"module": key, "files": [], "lines": 0, "bytes": 0})
        m["files"].append(f["path"])
        m["lines"] += f["lines"]
        m["bytes"] += f["bytes"]
    listed = sorted(modules.values(), key=lambda m: -m["lines"])
    ctx.log(f"inventory: {len(files)} files in {len(listed)} modules under {root}")
    return {
        "root": root.replace("\\", "/"), "file_count": len(files), "total_lines": sum(f["lines"] for f in files), "modules": listed,
        "largest_files": sorted(files, key=lambda f: -f["lines"])[:10], "_stats": {"in": len(files), "out": len(listed)},
    }
